### ISSUE TITLE
pkglistgen: integrate drop list generator

### ISSUE THREAD

**Issue by jberry-suse** [openSUSE/osc-plugin-factory#1375]
Per [poo#30225: enable drop list generator](https://progress.opensuse.org/issues/30225), it seems there are a few things that need to be reconciled.

- How to the `.solv` files in `trees/` (not in git, but on `packagelists` machine) relate to those in `create-drop-list`? From what I can surmise they represent the pre-published `standard` repo before it enters the _totest_ workflow. Meaning what is available from `https://api.opensuse.org/public/build/` instead of `http://download.opensuse.org/`.
- How are `do_create_droplist` and `do_dump_solv` intended to work together since they expect different file name formats?

  https://github.com/openSUSE/osc-plugin-factory/blob/5a472eb4006ba59ef5487eb3ab764a324fdae7b2/pkglistgen.py#L782-L784

  vs

  https://github.com/openSUSE/osc-plugin-factory/blob/5a472eb4006ba59ef5487eb3ab764a324fdae7b2/pkglistgen.py#L835-L844

  Which is something like:
  - `repo-openSUSE:Leap:15.0-standard-x86_64.solv`
  - `openSUSE-15.0-x86_64-Build109.3.solv`
- How does one generate the `repo-` (pre-released) files using `pkglistgen.py`?
- Is the `obsoletepackages.inc`, generated by `do_create_droplist`, to be placed into `000package-groups` and copied to `000product`? Or would it make sense to only keep in `000product` since it is generated just like the rest of solving process?
- The drop list is only to be done for target project and ports and not for stagings?
- How does one reconcile the `non-oss` and `oss` repos as the [current `package-lists` repo](https://github.com/openSUSE/package-lists/tree/master/create-drop-list/Leap:42.3) only has one `solv` file. Presumably that would mean `do_create_droplist` would also need to be able to handle multiple target projects to include `:NonFree` which it currently does not. Is this expected?
- Should the new code commit to the old repo, `package-lists`, the new `solv` files? It seems like it should be at least noted somewhere that the new code cannot generate `solv` files for < _Leap 15.0_. As such it clearly cannot generate a cache even if desired.

Overall, seems like a lot more to figure out and code than "enable drop list generator" would imply. Also yet another task that would be greatly aided by any level of detail about how all the parts come together to create a workflow. Even just a dump of commands to run in order for an example workflow. Trying to reverse engineer this from the two code-bases is not ideal especially when it appears what remains is not to simply wire up the two subcommands to be called as part of `do_update_and_solve`.

I have a local stab that abstracts the download baseurl and calls the commands, but definitely some details that need answers. @lnussel

**Comment by jberry-suse** [comment]
The current code also has [`$proj/*.solv` and `*.solv`](https://github.com/openSUSE/package-lists/blob/3feda24499bb91ad798b09f21759bd5d9cd73e8d/create-drop-list.sh#L9-L12) as input to creating drop list. This would seem to imply generating against all other distributions (except `Factory`) including beta releases of itself. This seems to make sense, but can you confirm? Also if this is the case, then the recent change for `Leap:42.3` to have its own directory means that logic needs to be a bit different going forward since it will need to include prior _Leap_ directories? Obviously, done in such a way as to avoid `Factory` for _Leap_.

**Comment by DimStar77** [comment]
I'd say 15.0 only should care for the RELEASED products plus Leap-15.0/*.solv; all betas/snapshots of earlier releases are not important since the final upgrade would have taken care of removing cruft already.

And I doubt we care for somebody updating from a random 42.3 snapshot to 15.0 betas

(there are some oold *.solv that can be removed, like 12.x, the various -Snapshot and -Beta of existing releases)

**Comment by lnussel** [comment]
Jimmy Berry wrote:
> - How to the `.solv` files in `trees/` (not in git, but on
> `packagelists` machine) relate to those in `create-drop-list`? From
> what I can surmise they represent the pre-published `standard` repo
> before it enters the _totest_ workflow. Meaning what is available from
> `https://api.opensuse.org/public/build/` instead of
> `http://download.opensuse.org/`.

Indeed. That is what the package list generator also uses to compute the
content of the DVD.

> - How are `do_create_droplist` and `do_dump_solv` intended to work
> together since they expect different file name formats?
> 
>    https://github.com/openSUSE/osc-plugin-factory/blob/5a472eb4006ba59ef5487eb3ab764a324fdae7b2/pkglistgen.py#L782-L784
> 
>    vs
> 
>    https://github.com/openSUSE/osc-plugin-factory/blob/5a472eb4006ba59ef5487eb3ab764a324fdae7b2/pkglistgen.py#L835-L844
> 
>    Which is something like:
>    - `repo-openSUSE:Leap:15.0-standard-x86_64.solv`
>    - `openSUSE-15.0-x86_64-Build109.3.solv`

The former is the current built wheres the latter is a published one. So `repo-openSUSE:Leap:15.0-standard-x86_64.solv` would be used in the `for prp in self.tool.repos` loop, whereas `openSUSE-15.0-x86_64-Build109.3.solv` would have to be passed as argument to appear in `oldsolv`.

> - How does one generate the `repo-` (pre-released files using `pkglistgen.py`?)

`do_update()` does that as part of `do_update_and_solve()`. That's
why I think it makes sense to integrate the drop list generator
into pkglistgen.

> - Is the `obsoletepackages.inc`, generated by `do_create_droplist`, to be placed into `000package-groups` and copied to `000product`? Or would it make sense to only keep in `000product` since it is generated just like the rest of solving process?

`000product`, yes

> - The drop list is only to be done for target project and ports and not for stagings?

I think so, yes.

> - How does one reconcile the `non-oss` and `oss` repos as the [current `package-lists` repo](https://github.com/openSUSE/package-lists/tree/master/create-drop-list/Leap:42.3) only has one `solv` file. Presumably that would mean `do_create_droplist` would also need to be able to handle multiple target projects to include `:NonFree` which it currently does not. Is this expected?

The old code merged the solve files from oss and non-oss: https://github.com/openSUSE/package-lists/blob/master/create-drop-list/create-solv.sh#L42
The non-oss repo is usually not added as separate product, so the
`openSUSE-Addon-NonOss-release` package doesn't get installed
Therefore the obsolete non-oss list is generated into the
`openSUSE-release` package too.

> - Should the new code commit to the old repo, `package-lists`, the
> new `solv` files? It seems like it should be at least noted
> somewhere that the new code cannot generate `solv` files for <
> _Leap 15.0_. As such it clearly cannot generate a cache even if
> desired.

Not sure what the best way is. Using git to store those binary files
isn't the best thing either anyways. Or maybe we can create separate
git repos for separate target projects, so the repo can be thrown
away when released. We don't need to support older releases than 15
anymore. They are done.

> Overall, seems like a lot more to figure out and code than "enable
> drop list generator" would imply. Also yet another task that would
> be greatly aided by any level of detail about how all the parts
> come together to create a workflow. Even just a dump of commands
> to run in order for an example workflow. Trying to reverse

Something like

`pkglistgen.py -o /path/to/15.0_archive dump_solv --overwrite http://download.opensuse.org/distribution/leap/15.0/repo/oss/`

to store and archive the currently published build

Then

`pkglistgen.py -r openSUSE:Leap:15.0/standard -a x86_64 create_droplist openSUSE-15.0-x86_64-Build109.3.solv /path/to/15.0_archive/*.solv /path/to/other_old_solv_files/*.solv`

cu
Ludwig

-- 
  (o_   Ludwig Nussel
  //\
  V_/_  http://www.suse.com/
SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard,
Graham Norton, HRB 21284 (AG Nürnberg)

**Comment by jberry-suse** [comment]
> I'd say 15.0 only should care for the RELEASED products plus Leap-15.0/*.solv; all betas/snapshots of earlier releases are not important since the final upgrade would have taken care of removing cruft already.

So the _latest_ build from `Leap:42.3` and all from self (`Leap:15.0`)?

> (there are some oold *.solv that can be removed, like 12.x, the various -Snapshot and -Beta of existing releases)

Should the undesirables simply be dropped and then the code will just check against all? or do we want to move the previous version into directories and only take last one from each to be consistent?

**Comment by lnussel** [comment]
Jimmy Berry wrote:
>> I'd say 15.0 only should care for the RELEASED products plus Leap-15.0/*.solv; all betas/snapshots of earlier releases are not important since the final upgrade would have taken care of removing cruft already.
> 
> So the _latest_ build from `Leap:42.3` and all from self (`Leap:15.0`)?

Yes, GM only.

>> (there are some oold *.solv that can be removed, like 12.x, the various -Snapshot and -Beta of existing releases)
> 
> Should the undesirables simply be dropped and then the code will just
> check against all? or do we want to move the previous version into
> directories and only take last one from each to be consistent?

Fine with me to throw them away.

cu
Ludwig

-- 
  (o_   Ludwig Nussel
  //\
  V_/_  http://www.suse.com/
SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard,
Graham Norton, HRB 21284 (AG Nürnberg)

**Comment by jberry-suse** [comment]
Well after a lot of work, I believe I have it working. The underlying code really was not designed with its use in mind and it is very clear by the hacking around I had to do in order to get things working. Really the pre-existing code should be heavily refactored, but I'll likely leave it as is as this should have been done originally.

I'll clean it up tomorrow and make granular commits, but for now the newly generated [obsoletepackages.inc](https://gist.githubusercontent.com/jberry-suse/2b2d5ea6275a632e07810e0d995bba4a/raw/83471541f502b3853d7f47adc53f1c04a1a56227/obsoletepackages.inc).

I spent the time to work out determining projects within the same family in order to walk backwards as far as we want. This should also allow us to nuke a ton of the crap in `leaper.py` that is hard-coded and requires updating every _Leap_ version.
